packages/core/src/browser/tree/tree-decorator.ts

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { injectable } from 'inversify';
+import { Event, Emitter } from '../../common/event';
+
+/**
+ * CSS styles for the tree decorators.
+ */
+export namespace DecoratorStyles {
+    export const CAPTION_PREFIX = 'theia-caption-prefix';
+    export const CAPTION_SUFFIX = 'theia-caption-suffix';
+    export const ICON_WRAPPER = 'theia-icon-wrapper';
+    export const DECORATOR_SIZE = 'theia-decorator-size';
+    export const TOP_RIGHT = 'theia-top-right';
+    export const BOTTOM_RIGHT = 'theia-bottom-right';
+    export const BOTTOM_LEFT = 'theia-bottom-left';
+    export const TOP_LEFT = 'theia-top-left';
+}
+
+/**
+ * Tree decorator service that can change the outlook and the style of the tree items within a widget.
+ */
+export const TreeDecorator = Symbol('TreeDecorator');
+export interface TreeDecorator {
+
+    /**
+     * The unique identifier of the decorator. Ought to be unique in the application.
+     */
+    readonly id: string;
+
+    /**
+     * Fired when the decoration has calculated all the decoration data. Keys are the identifiers of the tree items.
+     */
+    onDidChangeDecorations: Event<Map<string, DecorationData>>;
```

Comment:
Is it incremental? I mean if a decorator fires only for one id, what happens with other keys? Do they get removed or they preserved? I would be nice if it is incremental.

Reply:
No, it is not incremental.

Reply:
ok, I thought it can simplify the computation of decorators for problems, but the marker manager also does not have incremental notifications, so it is fine